### PR TITLE
Handle mask reduction intrinsic in VectorAPIExpansion

### PR DESCRIPTION
**Pull request by gita-omr**
- generate mTrueCount, mFirstTrue, mLastTrue, mToLong

Depends on https://github.com/eclipse/omr/pull/6783

### CODE REVIEW

**Comment by Akira1Saitoh** [comment]
I got the following trees when trying this PR. `mToLong` has the second argument (n7461n), which I think is incorrect.
```
 219932  n7270n   (  1)    mToLongVector128Int8 ()                                                     219932         [    0xffff5d34d020] bci=[43,35,766] rc=1 vc=7222 vn=- li=423 udi=- nc=2 flg=0x20
 219933  n7271n   (  1)      vcmpeqVector128Int8 ()                                                    219933         [    0xffff5d34d070] bci=[33,42,2048] rc=1 vc=7222 vn=- li=423 udi=- nc=2 flg=0x20
 219934  n7272n   (  1)        vloadi  <array-shadow>[#220  Shadow] [flags 0x80000611 0x0 ] ()         219934         [    0xffff5d34d0c0] bci=[7,32,3746] rc=1 vc=7222 vn=- li=423 udi=- nc=1 flg=0x20
 219935  n7273n   (  1)          aladd (X>=0 sharedMemory )                                            219935         [    0xffff5d34d110] bci=[-1,0,42] rc=1 vc=7222 vn=- li=423 udi=- nc=2 flg=0x100
 219936  n7464n   (  5)            ==>aRegLoad (in &GPR_0166) (X!=0 SeenRealReference sharedMemory )
 219937  n7275n   (  2)            lsub (highWordZero X>=0 cannotOverflow )                            219937         [    0xffff5d34d1b0] bci=[-1,0,42] rc=2 vc=7222 vn=- li=423 udi=- nc=2 flg=0x5100
 219938  n7459n   (  4)              ==>lRegLoad (in GPR_0161) (highWordZero X>=0 cannotOverflow SeenR 219938 ealReference )
 219939  n7277n   (  1)              lconst -16 (X!=0 X<=0 )                                           219939         [    0xffff5d34d250] bci=[-1,0,42] rc=1 vc=7222 vn=- li=423 udi=- nc=0 flg=0x204
 219940  n7278n   (  1)        vloadi  <array-shadow>[#220  Shadow] [flags 0x80000611 0x0 ] ()         219940         [    0xffff5d34d2a0] bci=[22,32,3746] rc=1 vc=7222 vn=- li=423 udi=- nc=1 flg=0x20
 219941  n7279n   (  1)          aladd (X>=0 sharedMemory )                                            219941         [    0xffff5d34d2f0] bci=[-1,0,42] rc=1 vc=7222 vn=- li=423 udi=- nc=2 flg=0x100
 219942  n7463n   (  5)            ==>aRegLoad (in &GPR_0165) (X!=0 SeenRealReference sharedMemory )
 219943  n7275n   (  2)            ==>lsub (highWordZero X>=0 cannotOverflow )
 219944  n7461n   (  5)      ==>aRegLoad (in &GPR_0163) (X!=0 SeenRealReference sharedMemory )
```

Perhaps the following code needs to be updated for MaskReduction.
https://github.com/eclipse-openj9/openj9/blob/53c0b4dbbab04bf4d7d0257cbbf941be991c4e32/runtime/compiler/optimizer/VectorAPIExpansion.cpp#L1636-L1639

**Comment by gita-omr** [comment]
Thanks @Akira1Saitoh for finding this issue https://github.com/eclipse-openj9/openj9/pull/16171#issuecomment-1291401285 

I think I fixed it. Please kindly try it out.

File: runtime/compiler/optimizer/VectorAPIExpansion.cpp [hunk 1]
```diff
@@ -1932,6 +1960,18 @@ TR::ILOpCodes TR_VectorAPIExpansion::ILOpcodeFromVectorAPIOpcode(int32_t vectorA
...
          }
       }
+   else if (opCodeType == MaskReduction)
+      {
+      switch (vectorAPIOpCode)
+         {
+         case VECTOR_OP_MASK_TRUECOUNT: return scalar ? TR::BadILOp : TR::ILOpCode::createVectorOpCode(TR::mTrueCount, vectorType);
+         case VECTOR_OP_MASK_FIRSTTRUE: return scalar ? TR::BadILOp : TR::ILOpCode::createVectorOpCode(TR::mFirstTrue, vectorType);
+         case VECTOR_OP_MASK_LASTTRUE:  return scalar ? TR::BadILOp : TR::ILOpCode::createVectorOpCode(TR::mLastTrue,  vectorType);
+         case VECTOR_OP_MASK_TOLONG:    return scalar ? TR::BadILOp : TR::ILOpCode::createVectorOpCode(TR::mToLongBits,    vectorType);
```
Comment (knn-k) [inline runtime/compiler/optimizer/VectorAPIExpansion.cpp:1970]: Please adjust the number of spaces before `vectorType`.

**Comment by Akira1Saitoh** [comment]
> Thanks @Akira1Saitoh for finding this issue [#16171 (comment)](https://github.com/eclipse-openj9/openj9/pull/16171#issuecomment-1291401285)
> 
> I think I fixed it. Please kindly try it out.

I confirmed that it is fixed now.

**Comment by Akira1Saitoh** [comment]
It seems that it is not vectorized if the mask is created by comparison of float/double vectors.
For example, the following code was not vectorized.
```
       for (int i = 0; i < a.length; i += SPECIES.length()) {
          FloatVector va = FloatVector.fromArray(SPECIES, a, i);
          FloatVector vb = FloatVector.fromArray(SPECIES, b, i);
          VectorMask<Float> vm = va.eq(vb);
          ints[i] = vm.trueCount();
       }
```

File: runtime/compiler/optimizer/VectorAPIExpansion.cpp [hunk 2]
```diff
@@ -1755,12 +1768,27 @@ TR::Node *TR_VectorAPIExpansion::fromBitsCoercedIntrinsicHandler(TR_VectorAPIExp
...
+   bool mask = (broadcastType == MODE_BITS_COERCED_LONG_TO_MASK);
+
    if (mode == checkScalarization)
-      return node;
+      return mask ? NULL : node;
 
    if (mode == checkVectorization)
       {
-      TR::ILOpCodes splatsOpCode = TR::ILOpCode::createVectorOpCode(TR::vsplats,
+      TR::ILOpCodes splatsOpCode = TR::ILOpCode::createVectorOpCode(mask ? TR::mLongBitsToMask: TR::vsplats,
```
Comment (Akira1Saitoh) [inline runtime/compiler/optimizer/VectorAPIExpansion.cpp:1791]: You need to do the same thing when the mode is `doVectorization`.

Comment (gita-omr) [reply runtime/compiler/optimizer/VectorAPIExpansion.cpp]: Fixed. Thanks!

**Comment by gita-omr** [comment]
Addressed more comments.

**Comment by gita-omr** [comment]
> It seems that it is not vectorized if the mask is created by comparison of float/double vectors. For example, the following code was not vectorized.
> 
> ```
>        for (int i = 0; i < a.length; i += SPECIES.length()) {
>           FloatVector va = FloatVector.fromArray(SPECIES, a, i);
>           FloatVector vb = FloatVector.fromArray(SPECIES, b, i);
>           VectorMask<Float> vm = va.eq(vb);
>           ints[i] = vm.trueCount();
>        }
> ```

This is a very interesting issue. Looks like even though the test is for `Float128` vectors, the element type of the mask is considered to be an `Int` (4 of them). The algorithm in VectorAPIExpansion checks the element type and the number of elements to determine the type of the vector or mask. Looks like that needs to be changed and the type needs to be derived from the first parameter of each intrinsic, which is a Class of the corresponding vector or mask. I think it would be better to do in a different PR.

**Comment by knn-k** [comment]
Jenkins test sanity all jdk19 depends eclipse/omr#6783

File: runtime/compiler/optimizer/VectorAPIExpansion.cpp (hunk 1 above)
Comment (gita-omr) [reply runtime/compiler/optimizer/VectorAPIExpansion.cpp]: Done. Also added minor improvement to tracing.

**Comment by knn-k** [comment]
Jenkins test sanity all jdk19 depends eclipse/omr#6783

**Comment by knn-k** [comment]
Jenkins test sanity alinux64,plinux,xlinux jdk19 depends eclipse/omr#6783

**Comment by knn-k** [comment]
Jenkins test sanity alinux64,amac jdk19 depends eclipse/omr#6783

**Comment by knn-k** [comment]
Jenkins test sanity amac,plinux,xlinux jdk19 depends https://github.com/eclipse/omr/pull/6783

**Comment by knn-k** [comment]
Jenkins test sanity amac,plinux,xlinux jdk19

**Review by knn-k — APPROVED** [review]

**Pull request merged**